components/carousel/index.jsx

```diff
@@ -0,0 +1,287 @@
+/* Copyright (c) 2015-present, salesforce.com, inc. All rights reserved */
+/* Licensed under BSD 3-Clause - see LICENSE.txt or git.io/sfdc-license */
+
+// # Carousel Component
+
+// Implements the [Carousel design pattern](https://www.lightningdesignsystem.com/components/carousel/) in React.
+
+import React from 'react';
+import PropTypes from 'prop-types';
+
+// ### shortid
+// [npmjs.com/package/shortid](https://www.npmjs.com/package/shortid)
+// shortid is a short, non-sequential, url-friendly, unique id generator
+import shortid from 'shortid';
+import '../../styles/carousel/carousel.css';
+
+import checkProps from './check-props';
+import componentDoc from './docs.json';
+
+import { CAROUSEL } from '../../utilities/constants';
+
+import CarouselIndicators from './private/carousel-indicators';
+import PreviousNextCarouselNavigator from './private/previous-next-carousel-navigator';
+import CarouselItem from './private/carousel-item';
+import AutoPlayButton from './private/auto-play-button';
+
+// ### Event Helpers
+import KEYS from '../../utilities/key-code';
+import EventUtil from '../../utilities/event';
+/* eslint-disable jsx-a11y/no-static-element-interactions */
+/**
+ * A carousel allows multiple pieces of featured content to occupy an allocated amount of space.
+ */
+class Carousel extends React.Component {
+	constructor(props) {
+		super(props);
+
+		const { items, itemsPerPanel } = this.props;
+
+		this.generatedId = shortid.generate();
```

Comment:
Id generation should to occur `onComponentMount` and not in the constructor to make sure there aren't collisions.